doc/mgmt/SONiC_PIM_SM_HLD.md

```diff
@@ -0,0 +1,1036 @@
+Following will be added in the syncd inorder to process the IPMC entry or entries AGE and DONTAGE data,
+
+1. Establish a connection with STATE_DB inorder to add or remove IPMC entries AGE and DONTAGE data.
+2. New notification handler (Notificationhandler::onIpmcEvent) API will be defined and registered with SAI to receive IPMC specific AGE and
+   DONTAGE notifications.
+   IPMC specific notification will be identified with the enum "SAI_SWITCH_NOTIFICATION_NAME_IPMC_EVENT" which will be added 
+   newly in SAI layer. 
+3. On receiving the IPMC entry/entries AGE or DONTAGE data, required validations will be done after 
+   deserializing the received data and then serialized unmodified data will be written to STATE_DB for further processing. A new API
+   will be defined to add or delete the data to/from the STATE_DB.
+
+##### 3.1.3.1.2 IPMC Entry deletion
+Whenever a IPMC entry is deleted for any reason, AGE and DONTAGE data belongs to that particular IPMC entry will be
+cleared from the STATE_DB.
+##### 3.1.3.1.3 SYNCD WARM restart
+To be updated based on the syncd warm restart feature understanding from broadcom.
+## 3.2 Datastructure Changes
+
+Struct mfpm_msg_hdr_t
+
+|    Data Type     | Field_name | Description |
+| -----------------| ---------- | ----------- |
+|  uint8_t         |  version   | Protocol version |
+|  mfpm_msg_type_t |  msg_type  | message type     |
+|  uint16_t        |  msg_len   | Length of the entire message including header length |
+
+Enum mfpm_msg_type_t
+
+|   Enum name    | Value |
+| --------------- | ---------- | 
+|   MFPM_MSG_TYPE_NONE  | 0  | 
+|   MFPM_MSG_TYPE_SET    |  1  | 
+|   MFPM_MSG_TYPE_GET | 2 |
+|   MFPM_MSG_TYPE_MAX | 3 |
+
+Enum mfpm_route_status_t
+
+|   Enum name    | Value |
+| ---------------| ---------- | 
+|   MFPM_ROUTE_NOT_FOUND  | 0  | 
+|   MFPM_ROUTE_ACTIVE   |  1  | 
+|   MFPM_ROUTE_INACTIVE | 2 |
+
+
+struct mfpm_route_info_t
+
+|   Data Type    | Field_name |  Description   |
+| -------------- | ---------- | -------------- |
+| unit32_t       | vrf_id     | VRF Id         |
+| char[INTERFACE_NAMSIZ]      | iif_name   | Incoming interface name |
+| struct ip_addr | src        | Source address of the route |
+| struct ip_addr | grp        | Group address of the route  |
+| mfpm_route_status_t         | entry_status | Entry aliveness status | 
+
+struct mfpm_kat_info_t
+
+|   Data Type    | Field_name |  Description   |
+| -------------- | ---------- | -------------- |
+| unit32_t       | vrf_id     | VRF Id         |
+| unsigned int   | keep_alive_time | Keep alive timer config to be applied in NPU |
+
+union mfpm_data_msg_t
+
+|   Data Type    | Field_name |  Description   |
+| -------------- | ---------- | -------------- |
+|  mfpm_kat_info_t | kat_info | Keep alive timer config to be set |
+|  mfpm_route_info_t | route_info | Info to get Route status from MFPM client |
+
+### 3.2.1 **SAI specific data structure and callback changes**
+
```

Comment:
Add about new attributes - SAI_SWITCH_ATTR_IPMC_EVENT_NOTIFY & SAI_SWITCH_ATTR_IPMC_AGING_TIME

More description below,
The IPMC registers a callback named sai_ipmc_event_notification_fn through the switch attribute SAI_SWITCH_ATTR_IPMC_EVENT_NOTIFY.
The SONiC then create/sets the IPMC entry with the optional attribute SAI_SWITCH_ATTR_IPMC_AGING_TIME.
ipmc_entry_attr[0].id = SAI_SWITCH_ATTR_IPMC_AGING_TIME;
ipmc_entry_attr[0].value.u32 = aging_time; // in seconds
If the attribute is not added then the IPMC entry does not age out. If the aging time is specified as 0, the IPMC entry does not age out. The aging time configured can be queried using a GET request of this attribute.
Once the IPMC entry ages out, SAI notifies SONiC using sai_ipmc_event_notification_fn callback. This callback provides events of type sai_ipmc_event_notification_data_t which has information about the specific IPMC entries that got aged out. The SONiC uses the IPMC entry information to delete these IPMC entries.

Reply:
@ashok-krishnegowda based on our earlier discussion SAI_SWITCH_ATTR_IPMC_AGING_TIME is global config. Please refer below discussion point.
SAI API sai_set_switch_attribute() will be used to set the IPMC aging time with the attribute mentioned below,
sai_get_switch_attribute() to get the configured value.
 
Below SAI switch attribute will be added for the IPMC age time setting, 
 
/**
     * @brief Dynamic IPMC entry aging time in seconds
     *
     * Zero means aging is disabled.
     *
     * @type sai_uint32_t
     * @flags CREATE_AND_SET
     * @default 0
     */
    SAI_SWITCH_ATTR_IPMC_AGING_TIME,